### ISSUE TITLE
MA0003 False Positive with Delegates

### ISSUE THREAD

**Issue by carldebilly** [meziantou/Meziantou.Analyzer#813]
### Version of the Meziantou.Analyzer NuGet package

2.0.170

### Rule Identifier

MA0003

### Target Framework

net 9.0

### C# Language version

C# 12

## Summary

Rule **MA0003** triggers false positives when calling delegates created by `LoggerMessage.Define<T>()`. The analyzer suggests naming parameters that cannot be named due to the delegate signature lacking parameter names.

## Problem Description

When using the high-performance logging pattern with `LoggerMessage.Define<T>()`, the generated delegates do not have named parameters in their signature. However, `MA0003` still triggers on calls to these delegates when passing `null` for the `Exception?` parameter, even though:

1. **This is the correct and documented way** to call LoggerMessage delegates
2. **Attempting to name the parameter results in a compile error** (CS1746)
3. **The only workaround is to suppress MA0003**, which defeats the purpose of the rule

## Code Example

```csharp
using Microsoft.Extensions.Logging; // v9.0.0 used here

public class LoggingExample
{
    private readonly ILogger<LoggingExample> _logger;

    // LoggerMessage.Define creates a delegate with this signature:
    // Action<ILogger, string, Exception?>
    // Note: No parameter names in the delegate signature!
    private static readonly Action<ILogger, string, Exception?> LogUserCreated =
        LoggerMessage.Define<string>(
            LogLevel.Information,
            new EventId(1001, nameof(CreateUser)),
            "User created: {UserName}");

    public void CreateUser(string userName)
    {
        // MA0003 FALSE POSITIVE: "Name the parameter to improve code readability"
        // This is the CORRECT and ONLY way to call LoggerMessage delegates!
        LogUserCreated(_logger, userName, null);

        // Attempting to fix MA0003 by naming the parameter fails at compile time:
        // LogUserCreated(_logger, userName, exception: null);
        // CS1746: The delegate 'Action<ILogger, string, Exception?>' does not have a parameter named 'exception'

        // It works by suppressing MA0003:
        #pragma warning disable MA0003
        LogUserCreated(_logger, userName, null);
        #pragma warning restore MA0003
    }
}
```

## Expected Behavior
- MA0003 should **not trigger** on calls to delegates created by `LoggerMessage.Define<T>()`
- The rule should recognize this is a special case where parameter naming is not possible
- Developers should be able to use LoggerMessage delegates without suppressions

## Suggested Fix

The analyzer should detect calls to delegates (like those created by `LoggerMessage.Define<T>()`) and exclude them from MA0003 rule evaluation. 

## Test Project

A complete reproduction project demonstrating this issue is available showing:
- MA0003 false positives on LoggerMessage.Define delegate calls
- Compile errors when attempting to name parameters
- Comparison with modern LoggerMessage source generators that work correctly

---

This false positive significantly impacts code quality enforcement in projects using LoggerMessage.Define, forcing developers to choose between high-performance logging and clean analyzer compliance.

### Reproduction Steps

(see before)

**Comment by meziantou** [comment]
First, it's possible to name an argument when calling a delegate

````c#
LogUserCreated(_logger, userName, arg3: null);
````

However, I don't think this is relevant for `Action` or `Func` delegates as argument names are meaningless. The rule should detect that and stop reporting it.

**Comment by meziantou** [comment]
Fixed by cfd5e8b37d0e43c9c46b19c07d19e2b24c2ee02a
